File: power_grid_model_c/power_grid_model_c/src/options.cpp
// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
//
// SPDX-License-Identifier: MPL-2.0

#define PGM_DLL_EXPORTS
#include "forward_declarations.hpp"

#include "power_grid_model_c/options.h"

#include "handle.hpp"
#include "input_sanitization.hpp"
#include "options.hpp"

namespace {
using namespace power_grid_model;

using power_grid_model_c::call_with_catch;
using power_grid_model_c::safe_ptr_get;
} // namespace

// options
PGM_Options* PGM_create_options(PGM_Handle* /* handle */) {
    return new PGM_Options{}; // NOSONAR(S5025)
}
void PGM_destroy_options(PGM_Options* opt) {
    delete opt; // NOSONAR(S5025)
}
void PGM_set_calculation_type(PGM_Handle* handle, PGM_Options* opt, PGM_Idx type) {
    call_with_catch(handle, [opt, type] { safe_ptr_get(opt).calculation_type = type; });
}
void PGM_set_calculation_method(PGM_Handle* handle, PGM_Options* opt, PGM_Idx method) {
    call_with_catch(handle, [opt, method] { safe_ptr_get(opt).calculation_method = method; });
}
void PGM_set_symmetric(PGM_Handle* handle, PGM_Options* opt, PGM_Idx sym) {
    call_with_catch(handle, [opt, sym] { safe_ptr_get(opt).symmetric = sym; });
}
void PGM_set_err_tol(PGM_Handle* handle, PGM_Options* opt, double err_tol) {
    call_with_catch(handle, [opt, err_tol] { safe_ptr_get(opt).err_tol = err_tol; });
}
void PGM_set_max_iter(PGM_Handle* handle, PGM_Options* opt, PGM_Idx max_iter) {
    call_with_catch(handle, [opt, max_iter] { safe_ptr_get(opt).max_iter = max_iter; });
}
void PGM_set_threading(PGM_Handle* handle, PGM_Options* opt, PGM_Idx threading) {
    call_with_catch(handle, [opt, threading] { safe_ptr_get(opt).threading = threading; });
}
void PGM_set_short_circuit_voltage_scaling(PGM_Handle* handle, PGM_Options* opt,
                                           PGM_Idx short_circuit_voltage_scaling) {
    call_with_catch(handle, [opt, short_circuit_voltage_scaling] {
        safe_ptr_get(opt).short_circuit_voltage_scaling = short_circuit_voltage_scaling;
    });
}
void PGM_set_tap_changing_strategy(PGM_Handle* handle, PGM_Options* opt, PGM_Idx tap_changing_strategy) {
    call_with_catch(handle,
                    [opt, tap_changing_strategy] { safe_ptr_get(opt).tap_changing_strategy = tap_changing_strategy; });
}
void PGM_set_experimental_features(PGM_Handle* handle, PGM_Options* opt, PGM_Idx experimental_features) {
    call_with_catch(handle,
                    [opt, experimental_features] { safe_ptr_get(opt).experimental_features = experimental_features; });
}

File: src/power_grid_model/_core/options.py
# SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
#
# SPDX-License-Identifier: MPL-2.0

"""
Option class
"""

from collections.abc import Callable
from typing import Any

from power_grid_model._core.power_grid_core import OptionsPtr, get_power_grid_core as get_pgc


class OptionSetter:
    """
    setter for options
    """

    _setter: Callable

    def __init__(self, setter):
        self._setter = setter

    def __set__(self, instance: "Options", value: Any):
        self._setter(instance.opt, value)

    def __get__(self, instance, owner):
        raise NotImplementedError("Cannot get option value!")


class Options:
    """
    Option class
    """

    _opt: OptionsPtr
    # option setter
    calculation_type = OptionSetter(get_pgc().set_calculation_type)
    calculation_method = OptionSetter(get_pgc().set_calculation_method)
    symmetric = OptionSetter(get_pgc().set_symmetric)
    error_tolerance = OptionSetter(get_pgc().set_err_tol)
    max_iterations = OptionSetter(get_pgc().set_max_iter)
    threading = OptionSetter(get_pgc().set_threading)
    tap_changing_strategy = OptionSetter(get_pgc().set_tap_changing_strategy)
    short_circuit_voltage_scaling = OptionSetter(get_pgc().set_short_circuit_voltage_scaling)
    experimental_features = OptionSetter(get_pgc().set_experimental_features)

    @property
    def opt(self) -> OptionsPtr:
        """

        Returns: Pointer to the option object

        """
        return self._opt

    def __new__(cls, *args, **kwargs):
        instance = super().__new__(cls, *args, **kwargs)
        instance._opt = get_pgc().create_options()
        return instance

    def __del__(self):
        get_pgc().destroy_options(self._opt)

    # not copyable
    def __copy__(self):
        raise NotImplementedError("Class not copyable")

    def __deepcopy__(self, memodict):
        raise NotImplementedError("class not copyable")
